File: js/array/09-two-sum/two-sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
module.exports = function twoSum(nums, target) {
  if (nums.length <= 1) {
    return [];
  }

  const map = new Map();

  for (let j = 0; j < nums.length; ++j) {
    const current = nums[j];
    const lookingFor = target - nums[j];

    if (!map.has(lookingFor)) {
      map.set(current, j);
    } else {
      return [map.get(lookingFor), j];
    }
  }

  return [];
};

File: java/src/main/java/com/github/mjeanroy/leetcode/easy/array/TwoSum.java
package com.github.mjeanroy.leetcode.easy.array;

import java.util.HashMap;
import java.util.Map;

public class TwoSum {

	public static int[] twoSum(int[] nums, int target) {
		if (nums == null || nums.length <= 1) {
			return new int[]{};
		}

		Map<Integer, Integer> indexes = new HashMap<>();
		for (int i = 0; i < nums.length; ++i) {
			int value = nums[i];
			int lookingFor = target - value;
			if (indexes.containsKey(lookingFor)) {
				return new int[] {indexes.get(lookingFor), i};
			}

			indexes.put(nums[i], i);
		}

		return new int[]{};
	}
}
